dimod/cyvariables.pyx

```diff
@@ -0,0 +1,263 @@
+# Copyright 2021 D-Wave Systems Inc.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+from numbers import Number
+
+from cpython.long cimport PyLong_Check
+from cpython.dict cimport PyDict_Size, PyDict_Contains
+from cpython.ref cimport PyObject
+
+from dimod.utilities import iter_safe_relabels
+
+cdef extern from "Python.h":
+    # not yet available as of cython 0.29.22
+    PyObject* PyDict_GetItemWithError(object p, object key) except? NULL
+
+
+cdef class cyVariables:
+    def __init__(self, object iterable=None):
+        self._index_to_label = dict()
+        self._label_to_index = dict()
+        self._stop = 0
+
+        if iterable is not None:
+            self._extend(iterable, permissive=True)
+
+    def __contains__(self, v):
+        return bool(self.count(v))
+
+    # todo: support slices
+    def __getitem__(self, Py_ssize_t idx):
+        if idx < 0:
+            idx = self._stop + idx
+
+        if idx >= self._stop:
+            raise IndexError('index out of range')
+
+        cdef object v
+        cdef object pyidx = idx
+        cdef PyObject* obj
+        if self._is_range():
+            v = pyidx
+        else:
+            # faster than self._index_to_label.get
+            obj = PyDict_GetItemWithError(self._index_to_label, pyidx)
+            if obj == NULL:
+                v = pyidx
+            else:
+                v = <object>obj  # correctly handles the ref count
+
+        return v
+
+    def __len__(self):
+        return self._stop
+
+    cpdef object _append(self, object v=None, bint permissive=False):
+        """Append a new variable.
+
+        Args:
+            v (hashable, optional):
+                Add a new variable. If `None`, a new label will be generated.
```

Comment:
Be nice to say something on how, something like,  "If `None`, generates a new label that is the lowest currently unused integer ..."

Reply:
I am not sure I want to "lock in" a particular behavior for backwards compatibility reasons.